ndb/src/google/cloud/ndb/_ports.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Ported implementations from the Google App Engine SDK.
+
+The following members have been brought in:
+
+* ``google.appengine.api.datastore_types.BlobKey``
+"""
+
+import functools
+
+from google.cloud.ndb import exceptions
+
+
+_MAX_STRING_LENGTH = 1500
+
+
+@functools.total_ordering
+class BlobKey:
+    """Key used to identify a blob in the blobstore.
+
+    .. note::
+
+        The blobstore was an early Google App Engine feature that later became
+        Google Cloud Storage.
+
+    This class is a simple wrapper a :class:`bytes` object. The bytes represent
+    a key used internally by the Blobstore API  to identify application blobs
+    (i.e. Google Cloud Storage objects). The key corresponds to the entity name
+    of the underlying object.
+
+    Args:
+        blob_key (Optional[bytes]): The key used for the blobstore.
+
+    Raises:
+        .BadValueError: If the ``blob_key`` exceeds 1500 bytes.
+        .BadValueError: If the ``blob_key`` is not :data:`None` or a
+            :class:`bytes` instance.
+    """
+
+    __slots__ = ("_blob_key",)
+
+    def __init__(self, blob_key):
+        if isinstance(blob_key, bytes):
+            if len(blob_key) > _MAX_STRING_LENGTH:
+                raise exceptions.BadValueError(
+                    "blob key must be under {:d} "
+                    "bytes.".format(_MAX_STRING_LENGTH)
+                )
+        elif blob_key is not None:
+            raise exceptions.BadValueError(
+                "blob key should be a string; received "
```

Comment:
"blob key should be *bytes*..."

Reply:
Yeah I copy-pasted this from `ValidateString()`, where both `str` and `unicode` were accepted in Python 2. 

I don't think it's necessary to preserve the original error message here because ISTM all callers are private and won't pass in the wrong type. In general do you think we need to worry about preserving the **text** of error messages or just the type?

Reply:
Just the type.

Reply:
The text should be helpful, obviously, which might mean adapting.